tests/system/action/user/test_email_client_server.py

```diff
@@ -0,0 +1,510 @@
+            {
+                "from": envelope.mail_from,
+                "to": envelope.rcpt_tos,
+                "data": content.decode("utf8", errors="replace"),
+            }
+        )
+        self.ret_status = "250 Message accepted for delivery"
+        return self.ret_status
+
+
+class AiosmtpdConnectionManager:
+    def __init__(self, handler: AIOHandler, auth: bool = False) -> None:
+        self.handler = handler
+        self.auth = auth
+
+    def __enter__(self) -> None:
+        auth_kwargs = {}
+        if self.auth:
+            auth_kwargs = {
+                "auth_required": True,
+                "authenticator": Authenticator(),
+            }
+
+        if EmailSettings.connection_security in [
+            ConnectionSecurity.STARTTLS,
+            ConnectionSecurity.SSLTLS,
+        ]:
+            ssl_context = ssl.create_default_context(ssl.Purpose.SERVER_AUTH)
+            ssl_context.load_default_certs(purpose=ssl.Purpose.SERVER_AUTH)
+            ssl_context.load_cert_chain("cert.pem", "key.pem")
+            ssl_context.check_hostname = False
+            ssl_context.verify_mode = ssl.VerifyMode.CERT_NONE
+            if EmailSettings.connection_security == ConnectionSecurity.SSLTLS:
+                # This is a hack: The aiosmtpd library does not issue AUTH in EHLO, if not starttls is used.
+                # For other methods (SSL/TLS and NONE) setting auth_require_tls allows AUTH. The intention is to
+                # allow AUTH before TLS (which is ok for NONE), but a hack for SSL/TLS since we have an
+                # encrypted connection
+                if self.auth:
+                    auth_kwargs["auth_require_tls"] = False
+                self.controller = Controller(
+                    self.handler,
+                    EmailSettings.host,
+                    EmailSettings.port,
+                    server_hostname="127.0.0.1",
+                    ssl_context=ssl_context,
+                    **auth_kwargs,  # type: ignore
+                )
+            else:
+                self.controller = Controller(
+                    self.handler,
+                    EmailSettings.host,
+                    EmailSettings.port,
+                    server_hostname="127.0.0.1",
+                    require_starttls=True,
+                    tls_context=ssl_context,
+                    **auth_kwargs,  # type: ignore
+                )
+        else:
+            if self.auth:
+                auth_kwargs["auth_require_tls"] = False
+            self.controller = Controller(
+                self.handler, EmailSettings.host, EmailSettings.port, **auth_kwargs  # type: ignore
+            )
+
+        self.controller.start()
+
+    def __exit__(
+        self,
+        exc_type: Optional[Type[BaseException]],
+        exc_value: Optional[BaseException],
+        exc_traceback: Optional[TracebackType],
+    ) -> None:
+        self.controller.stop(no_assert=True)
+
+
+class Authenticator:
```

Comment:
Please transform this to a function. The class itself is not needed.

Reply:
done